### PR TITLE
feat(findKey): add findKey

### PR DESCRIPTION
**Pull request by Na-hyunwoo**
Related issues: https://github.com/toss/es-toolkit/issues/715

Before implementing findKey in the compat layer, I preemptively worked on es-toolkit's findKey.

Hi!, @raon0211 

There is something that needs your opinion.

There was a point where I was focused while implementing it, and that was how strictly to keep the type. 

In the case of lodash's findKey, if an object is received as the first argument of findKey, the return value is a simple string. 

However, in the case of findKey that I implemented, the key value of the object is specified as the return value. 

Since this is not a compat layer, I think it is right to follow the principles of es-toolkit rather than lodash, so I need your opinion on that.

I believe that in order for TypeScript to be used meaningfully, it is right to strictly manage types. 

As a maintainer, please tell me your opinion!

I have one more question. 

After this work, I would like to implement the compat layer's findKey, findLastKey, and compat layer's findLastKey in that order. Is that okay?

findKey benchmark

### CODE REVIEW

File: docs/ja/reference/object/findKey.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# findKey
```
Comment (raon0211) [inline docs/ja/reference/object/findKey.md:1]: We actually generate documents in an automated way; you don't have to manually create them.

File: src/object/findKey.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *   'barney':  { 'age': 36, 'active': true },
+ *   'fred':    { 'age': 40, 'active': false },
+ *   'pebbles': { 'age': 1,  'active': true }
+ * };
+ * findKey(users, function(o) { return o.age < 40; }); => 'barney'
+ */
+export function findKey<T extends PropertyKey, K>(
+  obj: Record<T, K>,
+  predicate: (value: K, key: T, obj: Record<T, K>) => boolean
+): T | undefined {
```
Comment (raon0211) [inline src/object/findKey.ts:22]:
```suggestion
export function findKey<T extends Record<any, any>>(
  obj: T,
  predicate: (value: T[keyof T], key: keyof T, obj: T) => boolean
): keyof T | undefined {
```

I guess this type signature is more accurate than the original one. 

---

Note that `foo` is inferred as `number | string` in the first one, but is accurately inferred as `number` in the second one.

### Old interface

```ts
declare function findKey<T extends PropertyKey, K>(
  obj: Record<T, K>,
  predicate: (value: K, key: T, obj: Record<T, K>) => boolean
): T | undefined

findKey({ foo: 1, bar: '1' }, (val, key, obj) => {
  // foo is string | number here
  obj.foo
})
```

image

### New interface

```ts
declare function findKey<T extends Record<any, any>>(
  obj: T,
  predicate: (value: T[keyof T], key: keyof T, obj: T) => boolean
): T | undefined

findKey({ foo: 1, bar: '1' }, (val, key, obj) => {
  // foo is number here
  obj.foo
})
```

image

File: docs/ja/reference/object/findKey.md (hunk 1 above)
Comment (Na-hyunwoo) [reply docs/ja/reference/object/findKey.md:1]: @raon0211 Oh, I see... I guess I won't have to write docs from next time.

File: src/object/findKey.ts (hunk 2 above)
Comment (Na-hyunwoo) [reply src/object/findKey.ts:22]: @raon0211 

Wow, thank you for the kind review. 

In the case of the old interface, the type inference of the object's key and value is inferred independently, so accurate type inference is not possible. 

I will commit again, reflecting your review!!

**Comment by raon0211** [comment]
> I would like to implement the compat layer's findKey, findLastKey, and compat layer's findLastKey in that order. Is that okay?

This will help a lot! 😃

**Review by raon0211 — APPROVED** [review]
Thanks for your contribution! ☺️

**Review by raon0211 — DISMISSED** [review]

**Pull request merged**